Hello World example can't find mockito during unit tests

```
[ERROR] Test has error: unittest.loader._FailedTest.helloworld_tests
Traceback (most recent call last):
  File "/home/arcivanov/.pyenv/versions/3.8.2/lib/python3.8/unittest/case.py", line 60, in testPartExecutor
    yield
  File "/home/arcivanov/.pyenv/versions/3.8.2/lib/python3.8/unittest/case.py", line 676, in run
    self._callTestMethod(testMethod)
  File "/home/arcivanov/.pyenv/versions/3.8.2/lib/python3.8/unittest/case.py", line 633, in _callTestMethod
    method()
  File "/home/arcivanov/.pyenv/versions/3.8.2/lib/python3.8/unittest/loader.py", line 34, in testFailure
    raise self._exception
ImportError: Failed to import test module: helloworld_tests
Traceback (most recent call last):
  File "/home/arcivanov/.pyenv/versions/3.8.2/lib/python3.8/unittest/loader.py", line 154, in loadTestsFromName
    module = __import__(module_name)
  File "/home/arcivanov/Documents/src/arcivanov/helloworld/src/unittest/python/helloworld_tests.py", line 1, in <module>
    from mockito import mock, verify
ModuleNotFoundError: No module named 'mockito'


Traceback (most recent call last):
  File "/home/arcivanov/.pyenv/versions/3.8.2/envs/pyb-3.8.2/lib/python3.8/site-packages/pybuilder/cli.py", line 467, in main
    summary = reactor.build(
  File "/home/arcivanov/.pyenv/versions/3.8.2/envs/pyb-3.8.2/lib/python3.8/site-packages/pybuilder/reactor.py", line 235, in build
    execution_summary = self.build_execution_plan(tasks, execution_plan)
  File "/home/arcivanov/.pyenv/versions/3.8.2/envs/pyb-3.8.2/lib/python3.8/site-packages/pybuilder/reactor.py", line 270, in build_execution_plan
    task_execution_summaries = self.execution_manager.execute_execution_plan(
  File "/home/arcivanov/.pyenv/versions/3.8.2/envs/pyb-3.8.2/lib/python3.8/site-packages/pybuilder/execution.py", line 399, in execute_execution_plan
    summaries.append(self.execute_task(task, **kwargs))
  File "/home/arcivanov/.pyenv/versions/3.8.2/envs/pyb-3.8.2/lib/python3.8/site-packages/pybuilder/execution.py", line 346, in execute_task
    task.execute(self.logger, kwargs, _executable=_executable)
  File "/home/arcivanov/.pyenv/versions/3.8.2/envs/pyb-3.8.2/lib/python3.8/site-packages/pybuilder/execution.py", line 187, in execute
    executable.execute(argument_dict)
  File "/home/arcivanov/.pyenv/versions/3.8.2/envs/pyb-3.8.2/lib/python3.8/site-packages/pybuilder/execution.py", line 101, in execute
    self.callable(*arguments)
  File "/home/arcivanov/.pyenv/versions/3.8.2/envs/pyb-3.8.2/lib/python3.8/site-packages/pybuilder/plugins/python/unittest_plugin.py", line 67, in run_unit_tests
    run_tests(project, logger, reactor, "unittest", "unit tests")
  File "/home/arcivanov/.pyenv/versions/3.8.2/envs/pyb-3.8.2/lib/python3.8/site-packages/pybuilder/plugins/python/unittest_plugin.py", line 107, in run_tests
    raise BuildFailedException(msg)
pybuilder.errors.BuildFailedException: There were 1 error(s) and 0 failure(s) in unit tests
------------------------------------------------------------
BUILD FAILED - There were 1 error(s) and 0 failure(s) in unit tests (pybuilder/plugins/python/unittest_plugin.py:107)
------------------------------------------------------------
Build finished at 2020-04-08 16:09:45
Build took 10 seconds (10966 ms)
```

--- a/src/integrationtest/python/should_not_load_remote_classes_in_unittest_on_error_tests.py
+++ b/src/integrationtest/python/should_not_load_remote_classes_in_unittest_on_error_tests.py
@@ -1,0 +1,71 @@
+#   -*- coding: utf-8 -*-
+#
+#   This file is part of PyBuilder
+#
+#   Copyright 2011-2020 PyBuilder Team
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import sys
+import unittest
+
+from itest_support import IntegrationTestSupport
+from pybuilder.errors import BuildFailedException
+
+
+class Test(IntegrationTestSupport):
+    def test(self):
+        self.write_build_file("""
+from pybuilder.core import use_plugin, init
+
+use_plugin("python.core")
+use_plugin("python.unittest")
+
+name = "unittest-integration-test"
+default_task = "run_unit_tests"
+
+@init
+def init (project):
+    project.build_depends_on("mockito")
+""")
+        self.create_directory("src/main/python")
+        self.write_file("src/main/python/helloworld.py", r"""import sys
+
+def helloworld(out):
+    out.write("Hello world of Python\n")
+
+""")
+        self.create_directory("src/unittest/python")
+        self.write_file("src/unittest/python/helloworld_tests.py", r"""from mockito import mock, verify
+import unittest
+
+from helloworld import helloworld
+
+class HelloWorldTest(unittest.TestCase):
+    def test_should_issue_hello_world_message(self):
+        out = mock()
+
+        helloworld(out)
+
+        verify(out).write("Goodbye world of Python\n")
+""")
+
+        reactor = self.prepare_reactor()
+        try:
+            reactor.build()
+        except BuildFailedException:
+            self.assertTrue("mockito" not in sys.modules)
+
+
+if __name__ == "__main__":
+    unittest.main()

--- a/src/integrationtest/python/should_not_load_remote_classes_in_unittest_tests.py
+++ b/src/integrationtest/python/should_not_load_remote_classes_in_unittest_tests.py
@@ -1,0 +1,69 @@
+#   -*- coding: utf-8 -*-
+#
+#   This file is part of PyBuilder
+#
+#   Copyright 2011-2020 PyBuilder Team
+#
+#   Licensed under the Apache License, Version 2.0 (the "License");
+#   you may not use this file except in compliance with the License.
+#   You may obtain a copy of the License at
+#
+#       http://www.apache.org/licenses/LICENSE-2.0
+#
+#   Unless required by applicable law or agreed to in writing, software
+#   distributed under the License is distributed on an "AS IS" BASIS,
+#   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#   See the License for the specific language governing permissions and
+#   limitations under the License.
+
+import sys
+import unittest
+
+from itest_support import IntegrationTestSupport
+
+
+class Test(IntegrationTestSupport):
+    def test(self):
+        self.write_build_file("""
+from pybuilder.core import use_plugin, init
+
+use_plugin("python.core")
+use_plugin("python.unittest")
+
+name = "unittest-integration-test"
+default_task = "run_unit_tests"
+
+@init
+def init (project):
+    project.build_depends_on("mockito")
+""")
+        self.create_directory("src/main/python")
+        self.write_file("src/main/python/helloworld.py", r"""import sys
+
+def helloworld(out):
+    out.write("Hello world of Python\n")
+
+""")
+        self.create_directory("src/unittest/python")
+        self.write_file("src/unittest/python/helloworld_tests.py", r"""from mockito import mock, verify
+import unittest
+
+from helloworld import helloworld
+
+class HelloWorldTest(unittest.TestCase):
+    def test_should_issue_hello_world_message(self):
+        out = mock()
+
+        helloworld(out)
+
+        verify(out).write("Hello world of Python\n")
+""")
+
+        reactor = self.prepare_reactor()
+        reactor.build()
+
+        self.assertTrue("mockito" not in sys.modules)
+
+
+if __name__ == "__main__":
+    unittest.main()

--- a/src/main/python/pybuilder/plugins/python/remote_tools/__init__.py
+++ b/src/main/python/pybuilder/plugins/python/remote_tools/__init__.py
@@ -17,7 +17,7 @@
 #   limitations under the License.
 
 
-from pybuilder.remote import ctx, PipeShutdownError, RemoteObjectPipe, logger, log_to_stderr
+from pybuilder.remote import Process, PipeShutdownError, RemoteObjectPipe, logger, log_to_stderr
 
 __all__ = ["RemoteObjectPipe", "start_tool", "Tool", "PipeShutdownError", "logger"]
 
@@ -32,7 +32,7 @@ class Tool:
         pass
 
 
-def start_tool(tools, group=None, name=None, logging=None):
+def start_tool(pyenv, tools, group=None, name=None, logging=None):
     """
     Starts a tool process
     """
@@ -42,7 +42,8 @@ def start_tool(tools, group=None, name=None, logging=None):
         logger.setLevel(int(logging))
 
     pipe = RemoteObjectPipe.new_pipe()
-    proc = ctx.Process(group=group, target=_instrumented_tool, name=name, args=(tools, pipe))
+    proc = Process(pyenv, group=group, target=_instrumented_tool, name=name, args=(tools, pipe))
+
     try:
         proc.start()
     finally:

--- a/src/main/python/pybuilder/plugins/python/remote_tools/unittest_tool.py
+++ b/src/main/python/pybuilder/plugins/python/remote_tools/unittest_tool.py
@@ -42,6 +42,6 @@ class UnitTestTool(Tool):
         pipe.hide("unittest_tests")
 
 
-def start_unittest_tool(tools, test_modules, test_method_prefix, logging=0):
+def start_unittest_tool(pyenv, tools, test_modules, test_method_prefix, logging=0):
     tool = UnitTestTool(test_modules, test_method_prefix)
-    return start_tool(tools + [tool], name="unittest", logging=logging)
+    return start_tool(pyenv, tools + [tool], name="unittest", logging=logging)

--- a/src/main/python/pybuilder/plugins/python/unittest_plugin.py
+++ b/src/main/python/pybuilder/plugins/python/unittest_plugin.py
@@ -46,6 +46,7 @@ def init_test_source_directory(project):
     project.set_property_if_unset("unittest_module_glob", "*_tests")
     project.set_property_if_unset("unittest_file_suffix", None)  # deprecated, use unittest_module_glob.
     project.set_property_if_unset("unittest_test_method_prefix", None)
+    project.set_property_if_unset("unittest_python_env", "build")
     project.set_property_if_unset("unittest_runner", (
         lambda stream: __import__("xmlrunner").XMLTestRunner(output=project.expand_path("$dir_target/reports"),
                                                              stream=stream), "_make_result"))
@@ -59,6 +60,7 @@ def coverage_init(project, logger, reactor):
         project.get_property("_coverage_tasks").append(run_unit_tests)
         project.get_property("_coverage_config_prefixes")[run_unit_tests] = "ut"
         project.set_property("ut_coverage_name", "Python unit test")
+        # project.set_property("ut_coverage_python_env", project.get_property("unittest_python_env"))
 
 
 @task
@@ -88,9 +90,11 @@ def run_tests(project, logger, reactor, execution_prefix, execution_name):
     try:
         test_method_prefix = project.get_property("%s_test_method_prefix" % execution_prefix)
         runner_generator = project.get_property("%s_runner" % execution_prefix)
-        result, console_out = execute_tests_matching(reactor.tools, runner_generator, logger, test_dir, module_glob,
-                                                     test_method_prefix,
-                                                     project.get_property("remote_debug"))
+        result, console_out = execute_tests_matching(
+            reactor.python_env_registry[project.get_property("unittest_python_env")],
+            reactor.tools, runner_generator, logger, test_dir, module_glob,
+            test_method_prefix,
+            project.get_property("remote_debug"))
 
         if result.testsRun == 0:
             logger.warn("No %s executed.", execution_name)
@@ -119,12 +123,12 @@ def run_tests(project, logger, reactor, execution_prefix, execution_name):
         raise BuildFailedException("Unable to execute %s." % execution_name)
 
 
-def execute_tests(tools, runner_generator, logger, test_source, suffix, test_method_prefix=None, remote_debug=0):
-    return execute_tests_matching(tools, runner_generator, logger, test_source, "*{0}".format(suffix),
+def execute_tests(pyenv, tools, runner_generator, logger, test_source, suffix, test_method_prefix=None, remote_debug=0):
+    return execute_tests_matching(pyenv, tools, runner_generator, logger, test_source, "*{0}".format(suffix),
                                   test_method_prefix, remote_debug=remote_debug)
 
 
-def execute_tests_matching(tools, runner_generator, logger, test_source, file_glob, test_method_prefix=None,
+def execute_tests_matching(pyenv, tools, runner_generator, logger, test_source, file_glob, test_method_prefix=None,
                            remote_debug=0):
     output_log_file = StringIO()
     try:
@@ -134,7 +138,7 @@ def execute_tests_matching(tools, runner_generator, logger, test_source, file_gl
                                     _create_runner(runner_generator, output_log_file))
 
         try:
-            proc, pipe = start_unittest_tool(tools, test_modules, test_method_prefix, logging=remote_debug)
+            proc, pipe = start_unittest_tool(pyenv, tools, test_modules, test_method_prefix, logging=remote_debug)
             try:
                 pipe.register_remote(runner)
                 pipe.register_remote_type(unittest.result.TestResult)
@@ -149,8 +153,10 @@ def execute_tests_matching(tools, runner_generator, logger, test_source, file_gl
                     try:
                         proc.join()
                     finally:
-                        if sys.version_info[:2] >= (3, 7):
+                        try:
                             proc.close()
+                        except AttributeError:
+                            pass
 
             remote_closed_cause = pipe.remote_close_cause()
             if remote_closed_cause is not None:
@@ -200,7 +206,7 @@ def _instrument_result(logger, result):
     old_addFailure = result.addFailure
 
     def startTest(self, test):
-        self.test_names.append(test)
+        self.test_names.append(str(test))
         self.logger.debug("starting %s", test)
         old_startTest(test)
 

--- a/src/main/python/pybuilder/remote/__init__.py
+++ b/src/main/python/pybuilder/remote/__init__.py
@@ -21,6 +21,9 @@ from pybuilder.python_utils import patch_mp
 
 patch_mp()
 
+import sys  # noqa: E402
+from textwrap import dedent  # noqa: E402
+from os.path import normcase as nc, sep  # noqa: E402
 from io import BytesIO, StringIO  # noqa: E402
 from pickle import PickleError, Unpickler, UnpicklingError, HIGHEST_PROTOCOL  # noqa: E402
 from pickletools import dis  # noqa: E402
@@ -36,25 +39,211 @@ PICKLE_PROTOCOL_MAX = HIGHEST_PROTOCOL
 
 if PY2:
     ConnectionError = EnvironmentError
+    _compat_pickle = None
+else:
+    import _compat_pickle
 
 ctx = mp_get_context("spawn")
 ctx.allow_connection_pickling()
 logger = ctx.get_logger()
 
 __all__ = ["RemoteObjectPipe", "RemoteObjectError",
-           "ctx", "proxy_members", "PipeShutdownError", "log_to_stderr"]
+           "Process", "proxy_members", "PipeShutdownError", "log_to_stderr"]
+
+BUILTIN_MODULES = set(sys.builtin_module_names)
+
+
+class Process:
+    def __init__(self, pyenv, group=None, target=None, name=None, args=None):
+        self.pyenv = pyenv
+        self.proc = ctx.Process(group=group, target=target, name=name, args=args)
+
+    def start(self):
+        pyenv = self.pyenv
+        if PY2:
+            if IS_WIN:
+                from multiprocessing import forking as patch_module
+                tracker = None
+            else:
+                from billiard import spawn as patch_module
+                from billiard import semaphore_tracker as tracker
+        else:
+            from multiprocessing import spawn as patch_module
+            if not IS_WIN:
+                try:
+                    from multiprocessing import semaphore_tracker as tracker
+                except ImportError:
+                    from multiprocessing import resource_tracker as tracker
+            else:
+                tracker = None
+
+        # This is done to prevent polluting tracker's path with our path magic
+        if tracker:
+            tracker.getfd()
+
+        old_python_exe = patch_module._python_exe
+        patch_module._python_exe = pyenv.executable[0]  # pyenv's actual sys.executable
+
+        old_get_command_line = patch_module.get_command_line
+
+        def patched_get_command_line(**kwds):
+            cmd_line = old_get_command_line(**kwds)
+            result = list(pyenv.executable) + cmd_line[1:]
+            logger.debug("Starting process with %r", result)
+            return result
+
+        patch_module.get_command_line = patched_get_command_line
+
+        old_preparation_data = patch_module.get_preparation_data
+
+        def patched_preparation_data(name):
+            d = old_preparation_data(name)
+            sys_path = d["sys_path"]
+
+            # Python 2
+            if sys_path is sys.path:
+                sys_path = list(sys_path)
+                d["sys_path"] = sys_path
+
+            exec_prefix = nc(sys.exec_prefix) + sep
+
+            trailing_paths = []
+            for idx, path in enumerate(sys_path):
+                nc_path = nc(path)
+
+                if nc_path.startswith(exec_prefix):
+                    sys_path[idx] = pyenv.env_dir + sep + path[len(exec_prefix):]
+                    trailing_paths.append(path)
+
+            # Push current exec_prefix paths to the very end
+            sys_path.extend(trailing_paths)
+
+            logger.debug("Process sys.path will be: %r", sys_path)
+            return d
+
+        patch_module.get_preparation_data = patched_preparation_data
+        try:
+            return self.proc.start()
+        finally:
+            patch_module._python_exe = old_python_exe
+            patch_module.get_command_line = old_get_command_line
+            patch_module.get_preparation_data = old_preparation_data
+
+    def terminate(self):
+        return self.proc.terminate()
+
+    def kill(self):
+        return self.proc.kill()
+
+    def join(self, timeout=None):
+        return self.proc.join(timeout)
+
+    def is_alive(self):
+        return self.proc.is_alive()
+
+    def close(self):
+        return self.proc.close()
+
+    @property
+    def name(self):
+        return self.proc.name
+
+    @name.setter
+    def name(self, name):
+        self.proc.name = name
+
+    @property
+    def daemon(self):
+        return self.proc.daemon
+
+    @daemon.setter
+    def daemon(self, daemonic):
+        self.proc.daemon = daemonic
+
+    @property
+    def authkey(self):
+        return self.proc.authkey
+
+    @authkey.setter
+    def authkey(self, authkey):
+        self.proc.authkey = authkey
+
+    @property
+    def exitcode(self):
+        return self.proc.exitcode
+
+    @property
+    def ident(self):
+        return self.proc.ident
+
+    pid = ident
+
+    @property
+    def sentinel(self):
+        return self.proc.sentinel
+
+    def __repr__(self):
+        return repr(self.proc)
 
 
 class ProxyDef:
-    def __init__(self, remote_id, type_name, methods, fields):
+    def __init__(self, remote_id, module_name, type_name, is_type, methods, fields, spec_fields):
         self.remote_id = remote_id
+        self.module_name = module_name
         self.type_name = type_name
+        self.is_type = is_type
         self.methods = methods
         self.fields = fields
+        self.spec_fields = spec_fields
 
     def __repr__(self):
-        return "ProxyDef[remote_id=%r, type_name=%r, methods=%r, fields=%r]" % (self.remote_id, self.type_name,
-                                                                                self.methods, self.fields)
+        return "ProxyDef[remote_id=%r, module_name=%r, type_name=%r, is_type=%r," \
+               " methods=%r, fields=%r, spec_fields=%r]" % (self.remote_id,
+                                                            self.module_name,
+                                                            self.type_name,
+                                                            self.is_type,
+                                                            self.methods,
+                                                            self.fields,
+                                                            self.spec_fields)
+
+    def make_proxy_type(self):
+        """
+        Return a proxy type whose methods are given by `exposed`
+        """
+
+        remote_id = self.remote_id
+        methods = tuple(self.methods)
+        fields = tuple(self.fields)
+        dic = {}
+
+        body = ""
+        for meth in methods:
+            body += dedent("""
+            def %s(self, *args, **kwargs):
+                return self._BaseProxy__rop.call(%r, %r, args, kwargs)""" % (meth, remote_id, meth))
+
+        for field in fields:
+            body += dedent("""
+            def %s_getter(self):
+                return self._BaseProxy__rop.call_getattr(%r, %r)
+
+            def %s_setter(self, value):
+                return self._BaseProxy__rop.call_setattr(%r, %r, value)
+
+            %s = property(%s_getter, %s_setter)""" % (field, remote_id, field, field,
+                                                      remote_id, field, field, field, field))
+
+        exec(body, dic)
+        if self.is_type:
+            proxy_type = type(self.type_name, (_BaseProxyType, object), dic)
+        else:
+            proxy_type = type(self.type_name, (_BaseProxy, object), dic)
+
+        proxy_type.__module__ = self.module_name
+        proxy_type.__name__ = self.type_name
+        proxy_type.__methods__ = methods
+        proxy_type.__fields__ = fields
+        return proxy_type
 
 
 PICKLE_PID_TYPE_REMOTE_OBJ = 0
@@ -87,34 +276,41 @@ class RemoteObjectUnpickler(Unpickler, object):
 
     @classmethod
     def loads(cls, buf, _rop, *args, **kwargs):
-        file = BytesIO(buf)
-        return cls(file, *args, _rop=_rop, **kwargs).load()
+        f = BytesIO(buf)
+        return cls(f, *args, _rop=_rop, **kwargs).load()
 
 
-_PICKLE_SKIP_PID_CHECK_TYPES = {type(None), type, bool, int, float, str, bytes, bytearray, list, tuple, dict, set}
+_PICKLE_SKIP_PID_CHECK_TYPES = {type(None), bool, int, float, complex, str, bytes, bytearray, list, tuple, dict, set}
 
 
 class RemoteObjectPickler(ForkingPickler, object):
 
     def __init__(self, *args, **kwargs):
         self._rop = kwargs.pop("_rop")  # type: _RemoteObjectPipe
+        self._verify_types = set()
+        self.exc_persisted = []
         if PY2:
             kwargs["protocol"] = self._rop.pickle_version  # This is for full backwards compatibility with Python 2
             super(RemoteObjectPickler, self).__init__(*args, **kwargs)
         else:
             super(RemoteObjectPickler, self).__init__(args[0], self._rop.pickle_version, *args[1:])
 
-    def persistent_id(self, obj, exc_persisted=[]):  # Mutable default is intentional here
-        if type(obj) in _PICKLE_SKIP_PID_CHECK_TYPES:
+    def persistent_id(self, obj):  # Mutable default is intentional here
+        t_obj = obj if isinstance(obj, type) else type(obj)
+        if t_obj in _PICKLE_SKIP_PID_CHECK_TYPES:
             return None
 
-        # This is a weird trick. Since we need
+        exc_persisted = self.exc_persisted
+        # This is a weird trick.
         if obj in exc_persisted:
             exc_persisted.remove(obj)
             return None
 
         if isinstance(obj, _BaseProxy):
             return PICKLE_PID_TYPE_REMOTE_BACKREF, obj._BaseProxy__proxy_def.remote_id
+
+        if issubclass(t_obj, _BaseProxyType):
+            return PICKLE_PID_TYPE_REMOTE_BACKREF, t_obj._BaseProxy__proxy_def.remote_id
 
         rop = self._rop
         remote_obj = rop.get_remote(obj)
@@ -125,18 +321,23 @@ class RemoteObjectPickler(ForkingPickler, object):
             exc_persisted.append(obj)
             return PICKLE_PID_TYPE_REMOTE_EXC_TB, reduce_exception(obj)  # exception with traceback
 
+        if t_obj not in rop._verified_types:
+            if t_obj.__module__ not in BUILTIN_MODULES:
+                self._verify_types.add((t_obj.__module__, t_obj.__name__))
+
         return None
 
     @classmethod
     def dumps(cls, obj, _rop, *args, **kwargs):
         buf = BytesIO()
-        cls(buf, *args, _rop=_rop, **kwargs).dump(obj)
+        pickler = cls(buf, *args, _rop=_rop, **kwargs)
+        pickler.dump(obj)
         if logger.getEffectiveLevel() == 1:
             buf.seek(0)
             dis_log = StringIO()
             dis(buf, dis_log)
             logger.debug(dis_log.getvalue())
-        return buf.getvalue()
+        return buf.getvalue(), pickler._verify_types
 
 
 def rebuild_exception(ex, tb):
@@ -149,16 +350,14 @@ def reduce_exception(ex):
     return ex, getattr(ex, "__traceback__", None)
 
 
-#
-# Functions for finding the method names of an object
-#
-
 def proxy_members(obj, public=True, protected=True, add_callable=True, add_str=True):
     '''
     Return a list of names of methods of `obj` which do not start with '_'
     '''
     methods = []
     fields = []
+    spec_fields = {}
+
     for name in dir(obj):
         is_public = name[0] != '_'
         is_protected = name[0] == '_' and (len(name) == 1 or name[1] != '_')
@@ -173,7 +372,21 @@ def proxy_members(obj, public=True, protected=True, add_callable=True, add_str=T
                 methods.append(name)
             else:
                 fields.append(name)
-    return methods, fields
+
+    if isinstance(obj, BaseException):
+        if hasattr(obj, "__cause__"):
+            fields.append("__cause__")
+        if hasattr(obj, "__traceback__"):
+            fields.append("__traceback__")
+        if hasattr(obj, "__context__"):
+            fields.append("__context__")
+        if hasattr(obj, "__suppress_context__"):
+            fields.append("__suppress_context__")
+
+    if isinstance(obj, type) and not PY2:
+        spec_fields["__qualname__"] = obj.__qualname__
+
+    return methods, fields, spec_fields
 
 
 class RemoteObjectPipe:
@@ -212,6 +425,8 @@ class RemoteObjectPipe:
 
 
 def obj_id(obj):
+    if isinstance(obj, type):
+        return obj, id(obj)
     return type(obj), id(obj)
 
 
@@ -289,20 +504,28 @@ class _RemoteObjectSession:
     def hide(self, name):
         self._exposed_objs.pop(name)
 
-    def register_remote(self, obj, methods=None, fields=None):
+    def register_remote(self, obj, methods=None, fields=None, spec_fields=None):
         remote_id = self._remote_id
         self._remote_id = remote_id + 1
 
-        if methods is None or fields is None:
-            obj_methods, obj_fields = proxy_members(obj)
+        if methods is None or fields is None or spec_fields is None:
+            obj_methods, obj_fields, obj_spec_fields = proxy_members(obj)
             if methods is None:
                 methods = obj_methods
 
             if fields is None:
                 fields = obj_fields
 
-        obj_type = type(obj)
-        proxy = ProxyDef(remote_id, obj_type.__module__ + "." + obj_type.__name__, methods, fields)
+            if spec_fields is None:
+                spec_fields = obj_spec_fields
+
+        if isinstance(obj, type):
+            obj_type = obj
+        else:
+            obj_type = type(obj)
+
+        proxy = ProxyDef(remote_id, obj_type.__module__, obj_type.__name__, obj_type is obj,
+                         methods, fields, spec_fields)
 
         self._remote_objs[obj] = proxy
         self._remote_objs_ids[remote_id] = obj
@@ -342,15 +565,25 @@ class _RemoteObjectSession:
         if proxy_def:
             return proxy_def
 
-        t_obj = type(obj)
-        if t_obj in self._remote_types:
-            logger.debug("%r is type %r, which will register as remote", obj_id(obj), t_obj)
-            return self.register_remote(obj)
-
-        for remote_type in self._remote_types:
-            if isinstance(obj, remote_type):
-                logger.debug("%r is instance of type %r, which will register as remote", obj_id(obj), remote_type)
+        if not isinstance(obj, type):
+            t_obj = type(obj)
+            if t_obj in self._remote_types:
+                logger.debug("%r is instance of type %r, which will register as remote", obj_id(obj), t_obj)
                 return self.register_remote(obj)
+
+            for remote_type in self._remote_types:
+                if isinstance(obj, remote_type):
+                    logger.debug("%r is instance of type %r, which will register as remote", obj_id(obj), remote_type)
+                    return self.register_remote(obj)
+        else:
+            if obj in self._remote_types:
+                logger.debug("%r will register as remote", obj)
+                return self.register_remote(obj)
+
+            for remote_type in self._remote_types:
+                if issubclass(obj, remote_type):
+                    logger.debug("%r is subtype of type %r, which will register as remote", obj_id(obj), remote_type)
+                    return self.register_remote(obj)
 
 
 class RemoteObjectError(Exception):
@@ -363,39 +596,13 @@ class PipeShutdownError(RemoteObjectError):
 
 
 class _BaseProxy:
-    def __init__(self, __rom, __rop, __proxy_def):
-        self.__rom = __rom
+    def __init__(self, __rop, __proxy_def):
         self.__rop = __rop
         self.__proxy_def = __proxy_def
 
 
-def _make_proxy_type(proxy_def):
-    '''
-    Return a proxy type whose methods are given by `exposed`
-    '''
-    remote_id = proxy_def.remote_id
-    methods = tuple(proxy_def.methods)
-    fields = tuple(proxy_def.fields)
-    dic = {}
-
-    for meth in methods:
-        exec('''def %s(self, *args, **kwargs):
-        return self._BaseProxy__rop.call(%r, %r, args, kwargs)''' % (meth, remote_id, meth), dic)
-
-    for field in fields:
-        exec("""
-def %s_getter(self):
-    return self._BaseProxy__rop.call_getattr(%r, %r)
-
-def %s_setter(self, value):
-    return self._BaseProxy__rop.call_setattr(%r, %r, value)
-
-%s = property(%s_getter, %s_setter)""" % (field, remote_id, field, field, remote_id, field, field, field, field), dic)
-
-    proxy_type = type(proxy_def.type_name, (_BaseProxy, object), dic)
-    proxy_type.__methods__ = methods
-    proxy_type.__fields__ = fields
-    return proxy_type
+class _BaseProxyType:
+    pass
 
 
 ROP_CLOSE = 0
@@ -409,13 +616,16 @@ ROP_GET_EXPOSED_RESULT = 6
 ROP_GET_PROXY_DEF = 7
 ROP_GET_PROXY_DEF_RESULT = 8
 
-ROP_REMOTE_ACTION = 10
-ROP_REMOTE_ACTION_CALL = 11
-ROP_REMOTE_ACTION_GETATTR = 12
-ROP_REMOTE_ACTION_SETATTR = 13
-ROP_REMOTE_ACTION_REMOTE_ERROR = 14
-ROP_REMOTE_ACTION_RETURN = 15
-ROP_REMOTE_ACTION_EXCEPTION = 16
+ROP_VERIFY_TYPES = 9
+ROP_VERIFY_TYPES_RESULT = 10
+
+ROP_REMOTE_ACTION = 20
+ROP_REMOTE_ACTION_CALL = 21
+ROP_REMOTE_ACTION_GETATTR = 22
+ROP_REMOTE_ACTION_SETATTR = 23
+ROP_REMOTE_ACTION_REMOTE_ERROR = 24
+ROP_REMOTE_ACTION_RETURN = 25
+ROP_REMOTE_ACTION_EXCEPTION = 26
 
 
 class _RemoteObjectPipe(RemoteObjectPipe):
@@ -428,6 +638,7 @@ class _RemoteObjectPipe(RemoteObjectPipe):
         self._conn_c = self._conn_p = None
         self._remote_proxy_defs = {}
         self._remote_proxies = {}
+        self._verified_types = set()
 
         self.id = None
         self.conn = None  # type: ctx.Connection
@@ -456,8 +667,18 @@ class _RemoteObjectPipe(RemoteObjectPipe):
         remote_proxies = self._remote_proxies
         remote_proxy = remote_proxies.get(remote_id, None)
         if remote_proxy is None:
-            remote_proxy_type = _make_proxy_type(proxy_def)
-            remote_proxy = remote_proxy_type(self._ros, self, proxy_def)
+            remote_proxy_type = proxy_def.make_proxy_type()
+
+            if proxy_def.is_type:
+                remote_proxy = remote_proxy_type
+                remote_proxy._BaseProxy__rop = self
+                remote_proxy._BaseProxy__proxy_def = proxy_def
+            else:
+                remote_proxy = remote_proxy_type(self, proxy_def)
+
+            for k, v in proxy_def.spec_fields.items():
+                setattr(remote_proxy, k, v)
+
             remote_proxies[remote_id] = remote_proxy
             logger.debug("registered local proxy for remote ID %d: %r", remote_id, remote_proxy)
         return remote_proxy
@@ -537,6 +758,7 @@ class _RemoteObjectPipe(RemoteObjectPipe):
 
         self._remote_proxy_defs = {}
         self._remote_proxies = {}
+        self._verified_types = set()
         self._remote_close_cause = None
 
         # Not an error. We HAVE to make sure the first send uses minimally-supported Pickle version
@@ -573,13 +795,18 @@ class _RemoteObjectPipe(RemoteObjectPipe):
                 remote_name = data[2]
                 remote_name_action = data[3]
                 if remote_name_action == ROP_REMOTE_ACTION_CALL:
-                    func = getattr(obj, remote_name, None)
+                    call_args = data[4]
+                    call_kwargs = data[5]
+
+                    if isinstance(obj, type) and remote_name.startswith("__"):
+                        func = getattr(type(obj), remote_name, None)
+                        call_args = [obj] + list(call_args)
+                    else:
+                        func = getattr(obj, remote_name, None)
                     if not callable(func):
                         self._send_obj((ROP_REMOTE_ACTION_REMOTE_ERROR, remote_id,
                                         "%r is not a callable",
                                         (remote_name,)))
-                    call_args = data[4]
-                    call_kwargs = data[5]
 
                     logger.debug("calling %r.%r (remote ID %d) with args=%r, kwargs=%r",
                                  remote_name,
@@ -636,6 +863,42 @@ class _RemoteObjectPipe(RemoteObjectPipe):
                     else:
                         self._send_obj((ROP_REMOTE_ACTION_RETURN, return_val))
                     continue
+
+            if action_type == ROP_VERIFY_TYPES_RESULT:
+                new_verified = data[1]
+                proxy_types = data[2]
+                verified_types = self._verified_types
+                pickle_version = self.pickle_version
+
+                for module, name in new_verified:
+                    cls = find_class(module, name, pickle_version)
+                    verified_types.add(cls)
+
+                for module, name in proxy_types:
+                    cls = find_class(module, name, pickle_version)
+                    self.register_remote_type(cls)
+
+                return
+
+            if action_type == ROP_VERIFY_TYPES:
+                verify_types = data[1]
+                need_proxy = []
+                new_verified = []
+                if verify_types:
+                    verified_types = self._verified_types
+                    pickle_version = self.pickle_version
+
+                    for module_name in verify_types:
+                        module, name = module_name
+                        try:
+                            cls = find_class(module, name, pickle_version)
+                            verified_types.add(cls)
+                            new_verified.append(module_name)
+                        except Exception:
+                            need_proxy.append(module_name)
+
+                self._send_obj((ROP_VERIFY_TYPES_RESULT, new_verified, need_proxy))
+                continue
 
             if action_type == ROP_REMOTE_ACTION_REMOTE_ERROR:
                 remote_id = data[1]
@@ -720,16 +983,27 @@ class _RemoteObjectPipe(RemoteObjectPipe):
         if self._remote_close_cause is None:
             self._remote_close_cause = e
 
+    def _dump(self, obj):
+        while True:
+            buf, verify_types = RemoteObjectPickler.dumps(obj, self)
+            if verify_types:
+                self._send_obj((ROP_VERIFY_TYPES, verify_types))
+                self._recv()
+            else:
+                return buf
+
     if PY2 and IS_WIN:
         # Python 2 on Windows uses Python multiprocessing
 
         def _send_obj(self, obj):
             """Send a (picklable) object"""
-            logger.debug("sending %r", obj)
             if self.conn.closed:
                 raise OSError("handle is closed")
+
+            buf = self._dump(obj)
+            logger.debug("sending %r", obj)
             try:
-                self.conn.send_bytes(RemoteObjectPickler.dumps(obj, self))
+                self.conn.send_bytes(buf)
             except (ConnectionError, EOFError) as e:
                 logger.debug("failed to send %r", obj, exc_info=e)
                 try:
@@ -756,15 +1030,18 @@ class _RemoteObjectPipe(RemoteObjectPipe):
                     self._close()
             obj = RemoteObjectUnpickler.loads(buf, self)
             logger.debug("received %r", obj)
+
             return obj
     else:
         # Python 2 on Linux uses Billiard that is API-compatible with Python 3
         def _send_obj(self, obj):
             """Send a (picklable) object"""
-            logger.debug("sending %r", obj)
             self.conn._check_closed()
+
+            buf = self._dump(obj)
+            logger.debug("sending %r", obj)
             try:
-                self.conn._send_bytes(RemoteObjectPickler.dumps(obj, self))
+                self.conn._send_bytes(buf)
             except (ConnectionError, EOFError) as e:
                 logger.debug("failed to send %r", obj, exc_info=e)
                 try:
@@ -791,3 +1068,30 @@ class _RemoteObjectPipe(RemoteObjectPipe):
             obj = RemoteObjectUnpickler.loads(buf.getvalue(), self)
             logger.debug("received %r", obj)
             return obj
+
+
+def find_class(module, name, proto):
+    if proto < 3 and _compat_pickle:
+        if (module, name) in _compat_pickle.NAME_MAPPING:
+            module, name = _compat_pickle.NAME_MAPPING[(module, name)]
+        elif module in _compat_pickle.IMPORT_MAPPING:
+            module = _compat_pickle.IMPORT_MAPPING[module]
+    __import__(module, level=0)
+    if proto >= 4:
+        return _getattribute(sys.modules[module], name)[0]
+    else:
+        return getattr(sys.modules[module], name)
+
+
+def _getattribute(obj, name):
+    for subpath in name.split('.'):
+        if subpath == '<locals>':
+            raise AttributeError("Can't get local attribute {!r} on {!r}"
+                                 .format(name, obj))
+        try:
+            parent = obj
+            obj = getattr(obj, subpath)
+        except AttributeError:
+            raise AttributeError("Can't get attribute {!r} on {!r}"
+                                 .format(name, obj))
+    return obj, parent

--- a/src/main/python/pybuilder/scaffolding.py
+++ b/src/main/python/pybuilder/scaffolding.py
@@ -100,7 +100,7 @@ def update_project():
 
 def _create_pyproject_file():
     pyproject_contents = '''[build-system]
-requires = ["pybuilder"]
+requires = ["pybuilder>=0.12.0"]
 build-backend = "pybuilder.pep517"
 '''
     if os.path.exists("pyproject.toml"):

--- a/src/unittest/python/plugins/python/unittest_plugin_tests.py
+++ b/src/unittest/python/plugins/python/unittest_plugin_tests.py
@@ -72,7 +72,7 @@ class ExecuteTestsTests(TestCase):
         pipe = Mock()
         pipe.remote_close_cause.return_value = None
         tool.return_value = (Mock(), pipe)
-        execute_tests([], runner, self.mock_logger, "/path/to/test/sources", "_tests.py")
+        execute_tests(Mock(), [], runner, self.mock_logger, "/path/to/test/sources", "_tests.py")
 
         mock_discover_modules_matching.assert_called_with("/path/to/test/sources", "*_tests.py")
 
@@ -84,7 +84,7 @@ class ExecuteTestsTests(TestCase):
         pipe = Mock()
         pipe.remote_close_cause.return_value = None
         tool.return_value = (Mock(), pipe)
-        execute_tests_matching([], runner, self.mock_logger, "/path/to/test/sources", "*_tests.py")
+        execute_tests_matching(Mock(), [], runner, self.mock_logger, "/path/to/test/sources", "*_tests.py")
 
         mock_discover_modules_matching.assert_called_with("/path/to/test/sources", "*_tests.py")
 
@@ -100,7 +100,7 @@ class ExecuteTestsTests(TestCase):
         mock_unittest.defaultTestLoader.loadTestsFromNames.return_value = mock_tests
         runner.return_value.run.return_value = self.mock_result
 
-        actual, _ = execute_tests([], runner, self.mock_logger, "/path/to/test/sources", "_tests.py")
+        actual, _ = execute_tests(Mock(), [], runner, self.mock_logger, "/path/to/test/sources", "_tests.py")
 
         self.assertEqual(self.mock_result, actual)
 

